// ts-postgres/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorLevel = exports.DatabaseError = exports.ResultIterator = exports.Result = exports.DataType = exports.DataFormat = exports.SSLMode = exports.Client = void 0;
var client_1 = require("./client");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return client_1.Client; } });
Object.defineProperty(exports, "SSLMode", { enumerable: true, get: function () { return client_1.SSLMode; } });
var types_1 = require("./types");
Object.defineProperty(exports, "DataFormat", { enumerable: true, get: function () { return types_1.DataFormat; } });
Object.defineProperty(exports, "DataType", { enumerable: true, get: function () { return types_1.DataType; } });
__exportStar(require("./query"), exports);
var result_1 = require("./result");
Object.defineProperty(exports, "Result", { enumerable: true, get: function () { return result_1.Result; } });
Object.defineProperty(exports, "ResultIterator", { enumerable: true, get: function () { return result_1.ResultIterator; } });
var protocol_1 = require("./protocol");
Object.defineProperty(exports, "DatabaseError", { enumerable: true, get: function () { return protocol_1.DatabaseError; } });
Object.defineProperty(exports, "ErrorLevel", { enumerable: true, get: function () { return protocol_1.ErrorLevel; } });
