// ts-postgres/dist/src/result.js
"use strict";
/* eslint-disable @typescript-eslint/no-non-null-assertion */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var _ResultRowImpl_names, _ResultRowImpl_lookup;
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeResult = exports.ResultIterator = exports.Result = void 0;
function makeRecord(names, data) {
    const result = {};
    names.forEach((key, j) => result[key] = data[j]);
    return result;
}
class ResultRowImpl extends Array {
    constructor() {
        super(...arguments);
        _ResultRowImpl_names.set(this, void 0);
        _ResultRowImpl_lookup.set(this, void 0);
    }
    set(names, lookup, values) {
        __classPrivateFieldSet(this, _ResultRowImpl_names, names, "f");
        __classPrivateFieldSet(this, _ResultRowImpl_lookup, lookup, "f");
        this.push(...values);
    }
    /**
     * Return value for the provided column name.
     */
    get(name) {
        var _a;
        const i = (_a = __classPrivateFieldGet(this, _ResultRowImpl_lookup, "f")) === null || _a === void 0 ? void 0 : _a.get(name);
        if (i === undefined)
            throw new Error(`Invalid column name: ${String(name)}`);
        return this[i];
    }
    /**
     * Return an object mapping column names to values.
     */
    reify() {
        if (__classPrivateFieldGet(this, _ResultRowImpl_names, "f") === undefined)
            throw new Error('Column names not available');
        return makeRecord(__classPrivateFieldGet(this, _ResultRowImpl_names, "f"), this);
    }
}
_ResultRowImpl_names = new WeakMap(), _ResultRowImpl_lookup = new WeakMap();
/**
 * The awaited query result.
 *
 * Iterating over the result yields objects of the generic type parameter.
 */
class Result {
    constructor(names, rows, status) {
        this.names = names;
        this.rows = rows;
        this.status = status;
    }
    [Symbol.iterator]() {
        let i = 0;
        const rows = this.rows;
        const length = rows.length;
        const names = this.names;
        const shift = () => {
            const data = rows[i++];
            return makeRecord(names, data);
        };
        return {
            next: () => {
                if (i === length)
                    return { done: true, value: undefined };
                return { done: false, value: shift() };
            }
        };
    }
}
exports.Result = Result;
/**
 * The query result iterator.
 *
 * Iterating asynchronously yields objects of the generic type parameter.
 */
class ResultIterator extends Promise {
    constructor(names, data, executor) {
        super((resolve, reject) => {
            executor((status) => {
                const names = this.names || [];
                const data = this.data || [];
                const lookup = new Map();
                let i = 0;
                for (const name of names) {
                    lookup.set(name, i);
                    i++;
                }
                resolve(new Result(names, data.map(values => {
                    const row = new ResultRowImpl();
                    row.set(names, lookup, values);
                    return row;
                }), status));
            }, reject);
        });
        this.names = names;
        this.data = data;
        this.subscribers = [];
        this.done = false;
    }
    /**
     * Return the first item (if any) from the query results.
     */
    first() {
        var _a, e_1, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                for (var _d = true, _e = __asyncValues(this), _f; _f = yield _e.next(), _a = _f.done, !_a;) {
                    _c = _f.value;
                    _d = false;
                    try {
                        const row = _c;
                        return row;
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield _b.call(_e);
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    /**
     * Return the first item from the query results, or throw an error.
     */
    one() {
        var _a, e_2, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                for (var _d = true, _e = __asyncValues(this), _f; _f = yield _e.next(), _a = _f.done, !_a;) {
                    _c = _f.value;
                    _d = false;
                    try {
                        const row = _c;
                        return row;
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield _b.call(_e);
                }
                finally { if (e_2) throw e_2.error; }
            }
            throw new Error('Query returned an empty result');
        });
    }
    notify(done, status) {
        if (done)
            this.done = true;
        for (const subscriber of this.subscribers)
            subscriber(done, status);
        this.subscribers.length = 0;
    }
    [Symbol.asyncIterator]() {
        let i = 0;
        //const container = this.container;
        const shift = () => {
            const names = this.names;
            const values = this.data[i];
            i++;
            if (names === null) {
                throw new Error("Column name mapping missing.");
            }
            return makeRecord(names, values);
        };
        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        let error = null;
        this.catch((reason) => {
            error = new Error(reason);
        });
        return {
            next: () => __awaiter(this, void 0, void 0, function* () {
                if (error) {
                    throw error;
                }
                if (this.data.length <= i) {
                    if (this.done) {
                        return { done: true, value: undefined };
                    }
                    if (yield new Promise((resolve, reject) => {
                        this.subscribers.push((done, status) => {
                            if (typeof status !== 'undefined') {
                                reject(status);
                            }
                            else {
                                resolve(done);
                            }
                        });
                    })) {
                        return { done: true, value: undefined };
                    }
                }
                return { value: shift(), done: false };
            })
        };
    }
}
exports.ResultIterator = ResultIterator;
ResultIterator.prototype.constructor = Promise;
function makeResult(transform) {
    let dataHandler = null;
    const names = [];
    const rows = [];
    const p = new ResultIterator(names, rows, (resolve, reject) => {
        dataHandler = ((row) => {
            if (row === null || typeof row === 'string') {
                resolve(row);
                p.notify(true);
            }
            else if (Array.isArray(row)) {
                rows.push(row);
                p.notify(false);
            }
            else {
                reject(row);
                p.notify(true, row);
            }
        });
    });
    const nameHandler = (ns) => {
        names.length = 0;
        if (transform) {
            ns = ns.map(transform);
        }
        names.push(...ns);
    };
    return {
        iterator: p,
        dataHandler: dataHandler,
        nameHandler: nameHandler
    };
}
exports.makeResult = makeResult;
